File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/others/basedcyclic/RandomSequentialCyclic.java
package cujae.inf.ic.om.heuristic.assignment.others.basedcyclic;

import java.util.ArrayList;
import java.util.Random;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.factory.DistanceType;
import cujae.inf.ic.om.heuristic.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar clientes a los depósitos dn forma secuencial por depósitos escogiendo el depósito al azar*/
public class RandomSequentialCyclic extends ByNotUrgency {
	
	Random random = new Random();
	
	public static DistanceType distanceType = DistanceType.Real;
	private Solution solution = new Solution();	
	
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	private ArrayList<Integer> listIDDepots;
	private NumericMatrix costMatrix;
		
	public RandomSequentialCyclic() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
	
	@Override
	public void initialize() {
		listClusters = initializeClusters();
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());
		costMatrix = initializeCostMatrix(listCustomersToAssign, Problem.getProblem().getDepots(), distanceType);
	}	
	
	@Override
	public void assign() {
		int posElementMatrix = -1;
		
		int idCustomer = -1;
		double requestCustomer = 0.0;
		
		int posRDMDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
		
		RowCol rcBestElement = null;
		
		int countTry = 0;
		
		boolean isFull = false;
		
		int totalItems = listCustomersToAssign.size();
		int totalClusters = listIDDepots.size();
		
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()))
		{
			posRDMDepot = random.nextInt(listIDDepots.size());
			posElementMatrix = Problem.getProblem().getPosElement(listIDDepots.get(posRDMDepot));
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(listIDDepots.get(posRDMDepot)));
			
			posCluster = posRDMDepot;
		
			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();

				while(!isFull) 
				{
					rcBestElement = costMatrix.indexLowerValue(posElementMatrix, 0, posElementMatrix, (totalItems - 1));
					
					idCustomer = Problem.getProblem().getCustomers().get(rcBestElement.getCol()).getIDCustomer();
		        	requestCustomer = Problem.getProblem().getCustomers().get(rcBestElement.getCol()).getRequestCustomer();

					if(capacityDepot >= (requestCluster + requestCustomer))
					{	
						requestCluster += requestCustomer;
						
						listClusters.get(posCluster).setRequestCluster(requestCluster);
						listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);
					
						listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));

						costMatrix.fillValue(rcBestElement.getRow(), 0, rcBestElement.getRow(), (totalItems - 1), Double.POSITIVE_INFINITY);
						costMatrix.fillValue(0, rcBestElement.getCol(), (totalItems + totalClusters - 1), rcBestElement.getCol(), Double.POSITIVE_INFINITY);

						posElementMatrix = Problem.getProblem().getPosElement(listClusters.get(posCluster).getItemsOfCluster().get(listClusters.get(posCluster).getItemsOfCluster().size() - 1));
						
						if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
						{
							if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
								solution.getClusters().add(listClusters.remove(posCluster));
							else
								listClusters.remove(posCluster);
							
							listIDDepots.remove(posCluster);
							isFull = true;
						}
					} 
					else 
					{				
						costMatrix.setItem(rcBestElement.getRow(), rcBestElement.getCol(), Double.POSITIVE_INFINITY);
						countTry++;
						
						if((countTry >= listCustomersToAssign.size()))
						{
							if(!listClusters.get(posCluster).getItemsOfCluster().isEmpty())
								solution.getClusters().add(listClusters.get(posCluster));

							listClusters.remove(posCluster);
							
							listIDDepots.remove(posCluster);
							isFull = true;
						}
					}
				}
				isFull = false;
				countTry = 0;				
			}		
		}
	}

	@Override
	public Solution finish() {		
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		OSRMService.clearDistanceCache();
		
		return solution;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/others/basedcyclic/random_sequential_cyclic.py
import random
import numpy as np
from typing import List
from ...classical.by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class RandomSequentialCyclic(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_depots: List[Depot]
        self.list_id_depots: List[int]
        
    def to_clustering(self) -> Solution:
        self.initialize()
        self.assign()
        return self.finish()  

    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_depots = list(Problem.get_problem().get_depots())
        self.list_id_depots = list(Problem.get_problem().get_list_id_depots())
        
    def assign(self):    
        cost_matrix = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        is_full: bool = False
        
        while any(self.list_customers_to_assign) and self.list_clusters:
            pos_rdm_depot = random.randint(0, len(self.list_depots) - 1)
            current_depot = self.list_depots[pos_rdm_depot]
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(current_depot)
            current_capacity_depot = capacity_depot

            pos_cluster = pos_rdm_depot

            if pos_cluster != -1:
                cluster = self.list_clusters[pos_cluster]

                while not is_full and any(self.list_customers_to_assign):
                    request_cluster = cluster.get_request_cluster()
                    
                    cost_row = cost_matrix[((cluster.get_id_cluster() - len(self.list_customers_to_assign)) - 1), :] 
                    col = np.argmin(cost_row)
                    row = ((cluster.get_id_cluster() - len(self.list_customers_to_assign)) - 1)
                    
                    selected_customer = self.list_customers_to_assign[col]
                    id_customer = selected_customer.get_id_customer()
                    request_customer = selected_customer.get_request_customer()
                    
                    if current_capacity_depot >= request_customer:
                        current_capacity_depot -= request_customer
                        cluster.set_request_cluster(request_cluster + request_customer)
                        cluster.get_items_of_cluster().append(id_customer)
                        
                        self.list_customers_to_assign[col] = None
                        cost_matrix[:, col] = float('inf')
                        
                    else:
                        cost_matrix[row, col] = float('inf')
                        
                        if self.is_full_depot(self.list_customers_to_assign, request_cluster, capacity_depot):
                            if cluster.get_items_of_cluster():
                                self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                            else:
                                self.list_clusters.pop(pos_cluster)
                                
                            is_full = True
                            self.list_depots.pop(0)

                is_full = False

    def finish(self) -> Solution:
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)

        OSRMService.clear_distance_cache()
        
        return self.solution